### PR TITLE
feat(hogql): timezone support

### PR DESCRIPTION
**Pull request by mariusandra**
## Changes

- Changes HogQL function `toDateTime` to call `toDateTimeOrNull` instead of `parseDateTimeBestEffort`
- Adds `toUUID` as a helper when looking up events via uuid (alternative available now: toString(events.uuid), but that's slower... rather we'll suggest `events.uuid = toUUID('string')`)
- Refactors clickhouse identifier and string escaping and related utils into the unambiguously named `escape_sql.py`
- Always passes a timezone to all `toDateTime` and `now()` functions - mimics what I saw now used in insights
- Wraps all datetime column accesses with `toTimezone` calls - mimics what I saw now used in insights
- Adds support for printing constants of type `datetime` and `date`, with the correct timezone
- Adds better support/sanitization for printing integers and floats, including `-inf`, `inf` and `nan`

## How did you test this code?

Wrote tests.

### CODE REVIEW

File: posthog/models/event/events_query.py [hunk 1]
```diff
@@ -92,18 +92,18 @@ def run_events_query(
     # prevent accidentally future events from being visible by default
     before = query.before or (now() + timedelta(seconds=5)).isoformat()
     try:
-        timestamp = isoparse(before).strftime("%Y-%m-%d %H:%M:%S.%f")
+        parsed_date = isoparse(before)
```
Comment (thmsobrmlr) [inline posthog/models/event/events_query.py:95]: nit: couldn't follow why the change from timestamp to parsed date was necessary and am going to assume this was simply a small refactor

File: posthog/hogql/printer.py [hunk 2]
```diff
@@ -338,23 +343,15 @@ def visit_compare_operation(self, node: ast.CompareOperation):
...
-        elif isinstance(node.value, bool) and node.value is False:
-            return "false"
-        elif isinstance(node.value, int) or isinstance(node.value, float):
-            # :WATCH_OUT: isinstance(True, int) is True (!), so check for numbers lower down the chain
-            return str(node.value)
-        elif isinstance(node.value, str) or isinstance(node.value, list):
+        if self.dialect == "clickhouse" and (
+            isinstance(node.value, str) or isinstance(node.value, list) or isinstance(node.value, tuple)
+        ):
+            # inline the string in hogql, but use %(hogql_val_0)s in clichkouse
```
Comment (thmsobrmlr) [inline posthog/hogql/printer.py:349]:
```suggestion
            # inline the string in hogql, but use %(hogql_val_0)s in clickhouse
```

**Review by thmsobrmlr — APPROVED** [review]
TLDR: LGTM

I have nothing to object to. Could not follow with all the changes though. Most notably I didn't understand why we have a clickhouse and a hogql dialect.

File: posthog/models/event/events_query.py (hunk 1 above)
Comment (mariusandra) [reply posthog/models/event/events_query.py:95]: yep, a small refactor. now constants can also be `datetime`-s, so this leverages that, instead of requiring all "hogql users" to write these long timestamp strings `"%Y-%m-%d %H:%M:%S.%f`

File: posthog/hogql/database.py [hunk 3]
```diff
@@ -343,6 +344,16 @@ class Config:
     static_cohort_people: StaticCohortPeople = StaticCohortPeople()
     groups: Groups = Groups()
 
+    def __init__(self, timezone: Optional[str]):
+        super().__init__()
+        try:
+            self._timezone = pytz.timezone(timezone).zone if timezone else None
```
[the remark this answers is not in the document]
Comment (mariusandra) [reply posthog/hogql/database.py:350]: 👌

File: posthog/hogql/printer.py [hunk 4]
```diff
@@ -401,7 +398,22 @@ def visit_call(self, node: ast.Call):
...
+                    if len(args) != 0:
+                        raise ValueError(f"Function '{node.name}' expects no arguments.")
+                    args.append(self.visit(ast.Constant(value=self._get_timezone())))
+
+                if node.name == "toDateTime":
+                    if len(args) != 1:
+                        raise ValueError(f"Function '{node.name}' expects only one argument.")
+                    args.append(self.visit(ast.Constant(value=self._get_timezone())))
+
+                return f"{CLICKHOUSE_FUNCTIONS[node.name]}({', '.join(args)})"
```
[the remark this answers is not in the document]
Comment (mariusandra) [reply posthog/hogql/printer.py:414]: This has already gotten slightly refactored in https://github.com/PostHog/posthog/pull/14969

Not _exactly_ into such a dataclass, but more a simple tuple. I did move the timezone functions into a glorious `ADD_TIMEZONE_TO_FUNCTIONS` list, though some way to tag this when defining the function in constants.py might be even better 🤔 

Let's continue this function refactoring in the function refactor PR? :)
